packages/utils/src/PromiseUtils.ts

```diff
@@ -16,30 +16,22 @@ export class PromiseUtils {
     promise: Promise<T> | T,
     cleanup?: (val: T) => void | null
   ): CancelablePromise<T> {
-    let hasCanceled = false;
     let resolved: T | undefined;
+    let rejectFn: (val?: any) => void;
 
     const wrappedPromise = new Promise((resolve, reject) => {
-      Promise.resolve(promise).then(
-        val => {
-          if (hasCanceled) {
-            if (cleanup) {
-              cleanup(val);
-            }
-            reject(new CanceledPromiseError());
-          } else {
-            resolved = val;
-            resolve(val);
-          }
-        },
-        error =>
-          hasCanceled ? reject(new CanceledPromiseError()) : reject(error)
-      );
+      rejectFn = reject;
+      Promise.resolve(promise)
+        .then(val => {
+          resolved = val;
```

Comment:
We should still check if `hasCanceled` is true here, and run the cleanup if it has canceled.
This behaviour is used in a few places, where you can just call `.cancel()` on the promise and it will run the cleanup when it can.

An example is in IrisGridPanel: https://github.com/deephaven/web-client-ui/blob/main/packages/dashboard-core-plugins/src/panels/IrisGridPanel.tsx#L517

Reply:
I think I'm a bit confused about what is happening. I thought that this if statement covers it since if it's pending then nothing needs to be done. If it is fulfilled and rejected, then no cleanup necessary. If it is fulfilled and resolved, then resolved shouldn't be null so cleanup will run.
```
if (resolved != null && cleanup) {
  cleanup(resolved);
}
```

Reply:
This `.then` is after the underlying promise is fulfilled. The below block (`if (resolved != null && cleanup)`) is after the cancel is called.
I think there are four scenarios involving cancelation:
1) Original promise resolved, then cancel is called
2) Original promise rejected, then cancel is called
3) Cancel is called, then original promise is resolved
4) Cancel is called, then original promise is rejected

In both cases 1 and 3, when the original promise is resolved, the cleanup function should be called when cancel is called. Without checking `hasCanceled` within our `.then` here, that cleanup function wasn't getting called in scenario 3.

Reply:
Okay, I understand now. I mistakenly thought that reject and resolve statements act as return statements so they would stop the execution code once reject was called.